File: lib/Data/models/UserModel.dart
import 'package:hey_you/Data/TemporaryModifications.dart';
import 'package:cloud_firestore/cloud_firestore.dart';


class UserModel {
  String id;
  String email;
  String firstName;
  String lastName;
  String biography;
  Map<String, dynamic> quizAnswers;

  List<TemporaryModification> temporaryModifications;
  List<String> permanentModifications;
  Map<String, double> location;

  String currentMatch;

  int totalConnections;
  bool discoverable;
  int longestStreak;
  int currentStreak;
  DateTime lastMatch;
  DateTime? currentStreakTimer;

  String bucketName;

  UserModel({
    required this.email,
    required this.firstName,
    required this.lastName,
    required this.id,
    required this.biography,
    required this.quizAnswers,
    required this.temporaryModifications,
    required this.permanentModifications,
    required this.location,
    required this.currentMatch,
    required this.totalConnections,
    required this.discoverable,
    required this.longestStreak,
    required this.currentStreak,
    required this.lastMatch,
    required this.currentStreakTimer,
    required this.bucketName
  });

  UserModel.initial()
      : id = '',
        email = '',
        firstName = '',
        lastName = '',
        biography = '',
        quizAnswers = <String, int>{},
        temporaryModifications = [],
        permanentModifications = [],
        location = {},
        currentMatch = '',
        totalConnections = 0,
        discoverable = false,
        longestStreak = 0,
        currentStreak = 0,
        lastMatch = DateTime.now(),
        currentStreakTimer = DateTime.now(),
        bucketName = '';

  UserModel.fromJson(Map<String, dynamic> json)
      : id = json['id'] as String? ?? '',
        email = json['Email'] as String? ?? '',
        firstName = json['FirstName'] as String? ?? '',
        lastName = json['LastName'] as String? ?? '',
        biography = json['Biography'] as String? ?? '',
        quizAnswers = (json['QuestionAnswers'] is Map<String, dynamic>)
            ? Map<String, dynamic>.from(json['QuestionAnswers'])
            : {},
        temporaryModifications = (json['TemporaryModifications'] as List<dynamic>?)
            ?.map((e) => TemporaryModification.fromJson(e))
            .toList() ??
            [],
        permanentModifications = (json['PermanentModifications'] as List<dynamic>?)
            ?.map((e) => e as String)
            .toList() ??
            [],
        location = (json['Location'] != null)
            ? Map<String, double>.from(json['Location'] as Map)
            : {},
        currentMatch = json['CurrentMatch'] as String? ?? '',
        totalConnections = json['TotalConnections'] as int? ?? 0,
        discoverable = json['Discoverable'] as bool? ?? false,
        longestStreak = json['LongestStreak'] as int? ?? 0,
        currentStreak = json['CurrentStreak'] as int? ?? 0,
        lastMatch = (json['LastMatch'] != null)
            ? json['LastMatch'].toDate() ?? DateTime.now()
            : DateTime.now(),
        currentStreakTimer = (json['CurrentStreakTimer'] != null)
          ? (json['CurrentStreakTimer'] as Timestamp?)!.toDate()
          : DateTime.now(),
        bucketName = json['NearestBucket'] as String? ?? '';


  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'FirstName': firstName,
      'LastName': lastName,
      'Email': email,
      'Biography': biography,
      'QuestionAnswers': quizAnswers,
      'TemporaryModifications': temporaryModifications.map((e) => e.toJson()).toList(),
      'PermanentModifications': permanentModifications,
      'Location': location,
      'CurrentMatch': currentMatch,
      'TotalConnections': totalConnections,
      'Discoverable': discoverable,
      'LongestStreak': longestStreak,
      'CurrentStreak': currentStreak,
      'LastMatch': lastMatch.toIso8601String(),
      'CurrentStreakTimer': (currentStreakTimer == null) ? null : currentStreakTimer!.toIso8601String(),
    };
  }

  @override
  String toString() {
    return '''
      UserModel(
        id: $id,
        email: $email,
        name: $firstName $lastName,
        biography: ${biography.isEmpty ? '(none)' : biography},
        quizAnswers: $quizAnswers,
        temporaryModifications: [${temporaryModifications.map((mod) => mod.toString()).join(', ')}],
        permanentModifications: $permanentModifications,
        location: $location,
        currentMatch: $currentMatch,
        totalConnections: $totalConnections
      )''';
  }
}

File: lib/Python Backend/models/user_model.py
from pydantic import BaseModel
from typing import List, Dict, Any, Optional
from datetime import datetime, time, timezone
from models.models import Geolocation, TemporaryModification


class User(BaseModel):
    id: str
    email: str
    firstName: str
    lastName: str
    biography: str
    quizAnswers: Dict[str, Any]
    temporaryModifications: List[TemporaryModification]
    permanentModifications: List[str]
    location: Geolocation
    currentMatch: str
    previousConnections: List[str]
    totalConnections: int = 0
    discoverable: bool = False
    longestStreak: int = 0
    currentStreak: int = 0
    lastMatch: datetime = datetime.now()
    nearestBucket: str
    partition: str
    currentStreakTimer: Optional[datetime]
    llmScore: float = 0.0
    OCEANScore: float = 0.0
    llmCleanedJSON: Dict[str, Any] = {}
    llmEmbedding: Optional[List[float]] = None     # <-- NEW

    @classmethod
    def from_json(cls, data: Dict[str, Any]):
        return cls(
            id=data.get("id", ""),
            email=data.get("Email", ""),
            firstName=data.get("FirstName", ""),
            lastName=data.get("LastName", ""),
            biography=data.get("Biography", ""),
            quizAnswers=data.get("QuestionAnswers", {}),
            temporaryModifications=[
                TemporaryModification(
                    start=datetime.fromisoformat(tm["start"]),
                    modification=tm["modification"]
                ) for tm in data.get("TemporaryModifications", [])
            ],
            llmScore=float(data.get("LLMScore", data.get("llmScore", 0.0)) or 0.0),
            OCEANScore=float(data.get("OCEANScore", 0.0) or 0.0),
            llmCleanedJSON=data.get("llmCleanedJSON", {}),
            llmEmbedding=data.get("llmEmbedding"),              # <-- NEW
            permanentModifications=data.get("PermanentModifications", []),
            location=Geolocation.from_json(data.get("Location", {})),
            currentMatch=data.get("CurrentMatch") or "",
            previousConnections=data.get("PreviousConnections", []),
            totalConnections=data.get("TotalConnections", 0),
            discoverable=data.get("Discoverable", False),
            longestStreak=data.get("LongestStreak", 0),
            currentStreakTimer=data.get("CurrentStreakTimer", None),
            currentStreak=data.get("CurrentStreak", 0),
            lastMatch=data["LastMatch"] if "LastMatch" in data else datetime.now(),
            nearestBucket=data.get("NearestBucket", ""),
            partition=data.get("Partition", "default")
        )

    def to_json(self):
        return {
            "id": self.id,
            "FirstName": self.firstName,
            "LastName": self.lastName,
            "Email": self.email,
            "Biography": self.biography,
            "QuestionAnswers": self.quizAnswers,
            "LLMScore": self.llmScore,
            "OCEANScore": self.OCEANScore,
            "llmCleanedJSON": self.llmCleanedJSON,
            "llmEmbedding": self.llmEmbedding,      # <-- NEW
            "TemporaryModifications": [tm.to_dict() for tm in self.temporaryModifications],
            "PermanentModifications": self.permanentModifications,
            "Location": self.location.to_dict(),
            "CurrentMatch": self.currentMatch,
            "PreviousConnections": self.previousConnections,
            "TotalConnections": self.totalConnections,
            "Discoverable": self.discoverable,
            "LongestStreak": self.longestStreak,
            "CurrentStreakTimer": self.currentStreakTimer.isoformat(),
            "CurrentStreak": self.currentStreak,
            "LastMatch": self.lastMatch.isoformat(),
            "NearestBucket": self.nearestBucket,
            "Partition": self.partition
        }
